Having troubles cross compiling with RISCV64 and cmake

I'm trying to setup OpenBLAS to cross compiling from x86_64 to RISCV_ZVL256B.  I'm getting errors with cmake.  Can someone help with this?

export FC=riscv64-unknown-linux-gnu-gfortran
export CC=riscv64-unknown-linux-gnu-gcc
export CXX=riscv64-unknown-linux-gnu-g++
export AR=riscv64-unknown-linux-gnu-ar
export AS=riscv64-unknown-linux-gnu-gcc
export LD=riscv64-unknown-linux-gnu-ld
export HOSTCC=gcc
export HOSTFC=gfortran
export CXXFLAGS="-DTARGET=RISCV64_ZVL256B"
export CFLAGS="-DTARGET=RISCV64_ZVL256B"
export LD_LIBRARY_PATH=/home/ckerchner/tools/tt-riscv-toolchain-ae8a01f3/sysroot/lib

cmake -DTARGET=RISCV64_ZVL256B ..

  /home/ckerchner/tools/tt-riscv-toolchain-ae8a01f3/bin/riscv64-unknown-linux-gnu-gcc
  -DTARGET=RISCV64_ZVL256B -DFORCE_RISCV64_ZVL256B -march=native
  -DGEMM_MULTITHREAD_THRESHOLD=4 -DNO_AVX512 -DNO_PARALLEL_MAKE=0
  -I"/proj_sw/user_dev/ckerchner/OpenBLAS/build/getarch_build"
  -I"/proj_sw/user_dev/ckerchner/OpenBLAS"
  -I"/proj_sw/user_dev/ckerchner/OpenBLAS/build" -o
  CMakeFiles/cmTC_a895f.dir/getarch.c.o -c
  /proj_sw/user_dev/ckerchner/OpenBLAS/getarch.c

  riscv64-unknown-linux-gnu-gcc: error: '-march=native': ISA string must

This seems to be the offending area in system.cmake

# On x86_64 build getarch with march=native. This is required to detect AVX512 support in getarch.
if (X86_64 AND NOT (${CMAKE_C_COMPILER_ID} STREQUAL "PGI" OR ${CMAKE_C_COMPILER_ID} STREQUAL "NVHPC"))
  set(GETARCH_FLAGS "${GETARCH_FLAGS} -march=native")
endif ()

If I comment out that, I still get errors and unable to setup my OpenBLAS for RISCV64_ZVL256B.

I created my own cmake file

cat ../riscv64.cmake
# the name of the target operating system
set(CMAKE_SYSTEM_NAME RISCV64_ZVL256B)

# which compilers to use for C and C++
set(CMAKE_C_COMPILER   riscv64-unknown-linux-gnu-gcc)
set(CMAKE_CXX_COMPILER riscv64-unknown-linux-gnu-g++)

# where is the target environment located
set(CMAKE_FIND_ROOT_PATH  /home/ckerchner/tools/tt-riscv-toolchain-ae8a01f3/sysroot)

# adjust the default behavior of the FIND_XXX() commands:
# search programs in the host environment
set(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM NEVER)

# search headers and libraries in the target environment
set(CMAKE_FIND_ROOT_PATH_MODE_LIBRARY ONLY)
set(CMAKE_FIND_ROOT_PATH_MODE_INCLUDE ONLY)

I still get these errors

CMake Error at cmake/utils.cmake:59 (file):
  file STRINGS file
  "/proj_sw/user_dev/ckerchner/OpenBLAS/kernel/arm64/KERNEL.RISCV64_ZVL256B"
  cannot be read.
Call Stack (most recent call first):
  kernel/CMakeLists.txt:40 (ParseMakefileVars)
  kernel/CMakeLists.txt:1495 (build_core)

Changing flags helps:

export CXXFLAGS="-DRISCV64_ZVL256B"
export CFLAGS="-DRISCV64_ZVL256B"

make TARGET=RISCV64_ZVL256B BINARY=64 ARCH=riscv64 CC=riscv64-unknown-linux-gnu-gcc     AR=riscv64-unknown-linux-gnu-ar AS=riscv64-unknown-linux-gnu-gcc     LD=riscv64-unknown-linux-gnu-gcc FC=riscv64-unknown-linux-gnu-gfortran     HOSTCC=gcc HOSTFC=gfortran -j16

But these are the types of errors I'm seeing:

In file included from /proj_sw/user_dev/ckerchner/OpenBLAS/build/kernel/CMakeFiles/samin_k.c:9:
/proj_sw/user_dev/ckerchner/OpenBLAS/kernel/riscv64/amin_vector.c: In function 'samin_k':
/proj_sw/user_dev/ckerchner/OpenBLAS/kernel/riscv64/amin_vector.c:131:14: error: argument type 'vfloat32m1_t' requires the V ISA extension
  131 |         minf = EXTRACT_FLOAT(v_res);
In file included from /proj_sw/user_dev/ckerchner/OpenBLAS/build/kernel/CMakeFiles/samax_k.c:8:
/proj_sw/user_dev/ckerchner/OpenBLAS/kernel/riscv64/amax_vector.c: In function 'samax_k':
/proj_sw/user_dev/ckerchner/OpenBLAS/kernel/riscv64/amax_vector.c:122:14: error: argument type 'vfloat32m1_t' requires the V ISA extension
  122 |         maxf = EXTRACT_FLOAT(v_res);
/proj_sw/user_dev/ckerchner/OpenBLAS/kernel/riscv64/amin_vector.c:131:14: error: argument type 'vfloat32m1_t' requires the V ISA extension
/proj_sw/user_dev/ckerchner/OpenBLAS/kernel/riscv64/amax_vector.c:122:14: error: argument type 'vfloat32m1_t' requires the V ISA extension
In file included from /proj_sw/user_dev/ckerchner/OpenBLAS/build/kernel/CMakeFiles/smax_k.c:7:
/proj_sw/user_dev/ckerchner/OpenBLAS/kernel/riscv64/max_vector.c: In function 'smax_k':
/proj_sw/user_dev/ckerchner/OpenBLAS/kernel/riscv64/max_vector.c:127:14: error: argument type 'vfloat32m1_t' requires the V ISA extension
  127 |         maxf = EXTRACT_FLOAT(v_res);
/proj_sw/user_dev/ckerchner/OpenBLAS/kernel/riscv64/max_vector.c:127:14: error: argument type 'vfloat32m1_t' requires the V ISA extension

That helps get past almost all the errors.  Still have this:

/proj_sw/user_dev/ckerchner/OpenBLAS/build/kernel/CMakeFiles/sgemm_kernel.c:7:10: fatal error: /proj_sw/user_dev/ckerchner/OpenBLAS/kernel/riscv64/sgemm_kernel_2x2_zvl256b.c: No such file or directory
    7 | #include "/proj_sw/user_dev/ckerchner/OpenBLAS/kernel/riscv64/sgemm_kernel_2x2_zvl256b.c"

Looks good but I still get these during the link phase (linker is reporting 2 at a time):

caxpy_k.c:zaxpy_k.c:((.text.text+0x+0x1ea1e6)): undefined reference to `: undefined reference to `small_caxpy_kernelsmall_zaxpy_kernel'
':
cdotc_k.c:zdotc_k.c:((.text.text+0x+0x100100)): undefined reference to `: undefined reference to `small_cdot_kernelsmall_zdot_kernel'

--- a/kernel/riscv64/zaxpy_vector.c
+++ b/kernel/riscv64/zaxpy_vector.c
@@ -44,9 +44,9 @@ USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #endif
 
 #if !defined(DOUBLE)
-inline int  small_caxpy_kernel(BLASLONG n, BLASLONG dummy0, BLASLONG dummy1, FLOAT da_r, FLOAT da_i, FLOAT *x, BLASLONG inc_x, FLOAT *y, BLASLONG inc_y, FLOAT *dummy, BLASLONG dummy2)
+static inline int  small_caxpy_kernel(BLASLONG n, BLASLONG dummy0, BLASLONG dummy1, FLOAT da_r, FLOAT da_i, FLOAT *x, BLASLONG inc_x, FLOAT *y, BLASLONG inc_y, FLOAT *dummy, BLASLONG dummy2)
 #else
-inline int  small_zaxpy_kernel(BLASLONG n, BLASLONG dummy0, BLASLONG dummy1, FLOAT da_r, FLOAT da_i, FLOAT *x, BLASLONG inc_x, FLOAT *y, BLASLONG inc_y, FLOAT *dummy, BLASLONG dummy2)
+static inline int  small_zaxpy_kernel(BLASLONG n, BLASLONG dummy0, BLASLONG dummy1, FLOAT da_r, FLOAT da_i, FLOAT *x, BLASLONG inc_x, FLOAT *y, BLASLONG inc_y, FLOAT *dummy, BLASLONG dummy2)
 #endif
 {
 	BLASLONG i=0;

--- a/kernel/riscv64/zdot_vector.c
+++ b/kernel/riscv64/zdot_vector.c
@@ -69,9 +69,9 @@ USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 #endif
 
 #if !defined(DOUBLE)
- inline OPENBLAS_COMPLEX_FLOAT small_cdot_kernel(BLASLONG n, FLOAT *x, BLASLONG inc_x, FLOAT *y, BLASLONG inc_y)
+ static inline OPENBLAS_COMPLEX_FLOAT small_cdot_kernel(BLASLONG n, FLOAT *x, BLASLONG inc_x, FLOAT *y, BLASLONG inc_y)
 #else
-inline OPENBLAS_COMPLEX_FLOAT small_zdot_kernel(BLASLONG n, FLOAT *x, BLASLONG inc_x, FLOAT *y, BLASLONG inc_y)
+ static inline OPENBLAS_COMPLEX_FLOAT small_zdot_kernel(BLASLONG n, FLOAT *x, BLASLONG inc_x, FLOAT *y, BLASLONG inc_y)
 #endif
 {
 	BLASLONG i=0;

